src/Npgsql/TypeHandlers/DateTimeHandlers/DateHandler.cs

```diff
@@ -53,17 +67,35 @@ internal DateHandler(PostgresType postgresType, bool convertInfinityDateTime)
         /// <inheritdoc />
         public override DateTime Read(NpgsqlReadBuffer buf, int len, FieldDescription? fieldDescription = null)
         {
-            var npgsqlDate = ReadPsv(buf, len, fieldDescription);
-
-            if (npgsqlDate.IsFinite)
-                return (DateTime)npgsqlDate;
-            if (!_convertInfinityDateTime)
-                throw new InvalidCastException("Can't convert infinite date values to DateTime");
-            if (npgsqlDate.IsInfinity)
-                return DateTime.MaxValue;
-            return DateTime.MinValue;
+            const string infinityExceptionMessage = "Can't convert infinite date values to DateTime";
+            const string outOfRangeExceptionMessage = "Out of the range of DateTime (year must be between 1 and 9999)";
+
+            var postgresDate = buf.ReadInt32();
+
+            switch (postgresDate)
+            {
+            case int.MaxValue:
+                return _convertInfinityDateTime
+                    ? DateTime.MaxValue
+                    : throw new InvalidCastException(infinityExceptionMessage);
+            case int.MinValue:
+                return _convertInfinityDateTime
+                    ? DateTime.MinValue
+                    : throw new InvalidCastException(infinityExceptionMessage);
+            default:
+                try
+                {
+                    return FromPostgresDate(postgresDate);
```

Comment:
Any reason to not just inline the code here? It's a one-line expression method that's only used here... Same for the other handlers.